File: server/src/repositories/userRepository.ts
import { db } from "../db";
import { User, createUserDTO } from "../models/user";

const userCollection = db.collection("users");

export const createUser = async (user: createUserDTO): Promise<void> => {
  const now = new Date();
  try {
    const userDoc = await userCollection.doc(user.email).get();
    if (userDoc.exists) {
      throw new Error("User with this email already exists");
    }
    await userCollection.doc(user.email).set(
      {
        ...user,
        createdAt: user.createdAt || now,
        updatedAt: now,
      },
      { merge: true }
    );
  } catch (error) {
    console.error(`Error creating user with email ${user.email}:`, error);
    throw new Error("Failed to create user");
  }
};

export const getUserByEmail = async (email: string): Promise<User | null> => {
  try {
    const doc = await userCollection.doc(email).get();
    if (!doc.exists) {
      return null;
    }

    const data = doc.data();
    if (!data) return null;
    return {
      id: doc.id,
      ...data,
    } as User;
  } catch (error) {
    console.error(`Error fetching user by email ${email}:`, error);
    throw new Error("Failed to fetch user by email");
  }
};

File: server/internal/repository/user_repository.go
package repository

import (
	"context"
	"fmt"
	"mind-mate-server/graph/model"
	"mind-mate-server/internal/db"
	"time"

	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

func CreateUser(ctx context.Context, email string, firstName string, lastName string, dateOfBirth string, city string, country string, profession string, maritalStatus string) error {
	docRef := db.Client.Collection("users").Doc(email)

	// Check if already exists
	_, err := docRef.Get(ctx)
	if err == nil {
		return fmt.Errorf("user with this email already exists")
	}
	if status.Code(err) != codes.NotFound {
		return err
	}

	_, err = docRef.Set(ctx, map[string]interface{}{
		"email":         email,
		"firstName":     firstName,
		"lastName":      lastName,
		"dateOfBirth":   dateOfBirth,
		"city":          city,
		"country":       country,
		"profession":    profession,
		"maritalStatus": maritalStatus,
		"createdAt":     time.Now(),
		"updatedAt":     time.Now(),
	})
	return err
}

func GetUserByEmail(ctx context.Context, email string) (*model.User, error) {
	doc, err := db.Client.Collection("users").Doc(email).Get(ctx)
	if err != nil {
		if status.Code(err) == codes.NotFound {
			return nil, nil // Not found
		}
		return nil, err
	}

	data := doc.Data()
	return &model.User{
		Email:         getString(data, "email"),
		FirstName:     getString(data, "firstName"),
		LastName:      getString(data, "lastName"),
		DateOfBirth:   getStringPtr(data, "dateOfBirth"),
		City:          getStringPtr(data, "city"),
		Country:       getStringPtr(data, "country"),
		Profession:    getStringPtr(data, "profession"),
		MaritalStatus: getStringPtr(data, "maritalStatus"),
	}, nil
}

func getString(data map[string]interface{}, key string) string {
	if v, ok := data[key].(string); ok {
		return v
	}
	return ""
}

func getStringPtr(data map[string]interface{}, key string) *string {
	if v, ok := data[key].(string); ok {
		return &v
	}
	return nil
}
